### PR TITLE
[REVIEW] Refactor Index class, add CategoricalIndex

### PR DESCRIPTION
**Pull request by kkraus14**
This is a precursor to properly tackling #666 

Still need to add tests, but all current pytests should be passing

### CODE REVIEW

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
This isn't as much of a refactor as the Index class really needs, but it cleans some things up a bit and adds proper support for Categorical dtype indexes. It should also unblock allowing me to leverage pyarrow a lot more heavily for #666.

File: python/cudf/dataframe/index.py [hunk 1]
```diff
@@ -393,4 +382,86 @@ def get_dt_field(self, field):
...
+    * pandas.Categorical
+
+    Returns
+    -------
+    result : subclass of Index
+        - CategoricalIndex for Categorical input.
+        - DatetimeIndex for Datetime input.
+        - GenericIndex for all other inputs.
+    """
+    # This function should probably be moved to Index.__new__
```
Comment (dantegd) [inline python/cudf/dataframe/index.py]: Any particular reason why its not in `Index.__new__` yet?

File: python/cudf/tests/test_index.py [hunk 2]
```diff
@@ -98,3 +101,41 @@ def test_index_immutable():
...
     with pytest.raises(TypeError):
         gi[1] = 5
+
+
+def test_categorical_index():
+    pdf = pd.DataFrame()
+    pdf['a'] = [1, 2, 3]
+    pdf['index'] = pd.Categorical(['a', 'b', 'c'])
+    pdf = pdf.set_index('index')
+    gdf = DataFrame.from_pandas(pdf)
```
Comment (dantegd) [inline python/cudf/tests/test_index.py]: Wouldn't checking a dataframe/series with categorical index without being build with `from_pandas` be a good idea as well?

**Comment by kkraus14** [comment]
rerun tests

File: python/cudf/dataframe/index.py (hunk 1 above)
Comment (kkraus14) [reply python/cudf/dataframe/index.py]: I kinda followed the pattern of `columnops.as_column` for now, no particularly strong reason, but I'd like to hold off on further refactoring until 0.6+.

File: python/cudf/tests/test_index.py (hunk 2 above)
Comment (kkraus14) [reply python/cudf/tests/test_index.py]: Good point, will extend the tests.

File: python/cudf/dataframe/index.py (hunk 1 above)
Comment (dantegd) [reply python/cudf/dataframe/index.py]: fine by me, I was just curious

File: python/cudf/dataframe/dataframe.py [hunk 3]
```diff
@@ -491,7 +491,7 @@ def set_index(self, index):
             return df.set_index(self[index])
         # Otherwise
         else:
-            index = index if isinstance(index, Index) else GenericIndex(index)
+            index = index if isinstance(index, Index) else as_index(index)
```
Comment (thomcom) [inline python/cudf/dataframe/dataframe.py]: 👍

File: python/cudf/dataframe/index.py [hunk 4]
```diff
@@ -246,11 +245,11 @@ class GenericIndex(Index):
     _values: A Column object
     name: A string
     """
-    def __new__(self, values, name=None):
-        from .series import Series
-
+    def __init__(self, values, name=None):
```
Comment (thomcom) [inline python/cudf/dataframe/index.py]: 👍

File: python/cudf/dataframe/index.py [hunk 5]
```diff
@@ -393,4 +382,86 @@ def get_dt_field(self, field):
                                      mask=out_column.mask,
                                      null_count=out_column.null_count,
                                      dtype=out_column.dtype)
-        return GenericIndex(out_column)
+        return as_index(out_column)
+
+
+class CategoricalIndex(GenericIndex):
+    """An categorical of orderable values that represent the indices of another
```
Comment (thomcom) [inline python/cudf/dataframe/index.py]: super nitpick: "A categorical..."?

Comment (kkraus14) [reply python/cudf/dataframe/index.py]: It's a Pandas term, *shrug*

File: python/cudf/dataframe/index.py (hunk 1 above)
Comment (thomcom) [reply python/cudf/dataframe/index.py]: 👍

**Pull request merged**